### PR TITLE
Add GP Caching

### PR DESCRIPTION
**Pull request by ben-kaufman**

### CODE REVIEW

**Comment by orenyodfat** [comment]
looks ok 
please add tests

File: src/utils.ts [hunk 1]
```diff
@@ -1,16 +1,63 @@
...
   }
   for (let j = 0; j < b.length; j++) {
-    out[a.length + j] = b[j]
+    out[a.length + j] = b[j];
   }
-  return out as ByteArray
+  return out as ByteArray;
 }
 
+export function updateRedemption(
```
Comment (tsuberim) [inline src/utils.ts]: I would move this into `GenesisProtocol.ts`

**Comment by ben-kaufman** [comment]
Working on tests...

File: test/integration/GenesisProtocol.spec.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      let txs = [];
+      txs.push(await daoToken.methods.mint(accounts[0].address, "100").send());
+      txs.push(await daoToken.methods.mint(accounts[1].address, "100").send());
+      txs.push(await reputation.methods.mint(accounts[0].address, "60").send());
+      txs.push(await reputation.methods.mint(accounts[1].address, "40").send());
+      txs.push(
+        await genesisProtocolCallbacks.methods.setParameters(
+          [50, 60, 60, 1, 1, 0, 0, 60, 1, 10, 10, 80, 15, 10],
+          nullAddress
+        )
```
Comment (orenyodfat) [inline test/integration/GenesisProtocol.spec.ts]: .send() ?

Comment (ben-kaufman) [reply test/integration/GenesisProtocol.spec.ts]: Yes! I didn't notice I forgot that there, web3 1.0 :(

File: test/integration/GenesisProtocol.spec.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        await genesisProtocolCallbacks.methods.propose(
+          2,
+          paramsHash,
+          genesisProtocolCallbacks.options.address,
+          nullAddress,
+          nullAddress
+        )
+      );
+
+      txs.push(await genesisProtocol.methods.stake(proposalID, 0, 20));
```
Comment (tsuberim) [inline test/integration/GenesisProtocol.spec.ts]: missing `send`

File: test/integration/GenesisProtocol.spec.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        .call();
+
+      txs.push(
+        await genesisProtocolCallbacks.methods.propose(
+          2,
+          paramsHash,
+          genesisProtocolCallbacks.options.address,
+          nullAddress,
+          nullAddress
+        )
```
Comment (tsuberim) [inline test/integration/GenesisProtocol.spec.ts]: missing `send`

**Comment by orenyodfat** [comment]
added executionState and state to the GenesisProtocolProposal.
Please note that this part is comment in the gp mapping due to an issue with mapping uint8 vars (executionState and state ) .
See https://github.com/graphprotocol/graph-node/issues/569

File: ops/index.js [hunk 5]
```diff
@@ -85,13 +100,18 @@ async function migrate(web3) {
   const Token = new web3.eth.Contract(DAOToken.abi, undefined, opts);
   const token = await Token.deploy({
     data: DAOToken.bytecode,
-    arguments: ["TEST","TST",1000000000]
-    }).send();
+    arguments: ["TEST", "TST", 1000000000]
+  }).send();
+
+  const gpToken = await Token.deploy({
```
Comment (orenyodfat) [inline ops/index.js]: why gpToken ? can we just use token for gp ?

Comment (ben-kaufman) [reply ops/index.js]: No, it have the same problem as with using the same reputation. It makes the DAOToken test fail

Comment (orenyodfat) [reply ops/index.js]: so just deploy that locally at the GP tests

Comment (ben-kaufman) [reply ops/index.js]: I can't as the GP requires that in its constructor

**Comment by ben-kaufman** [comment]
@orenyodfat @tsuberim fixed the issues, can you review the PR?

**Comment by orenyodfat** [comment]
Please update proposal state on Vote and Stake.

**Comment by ben-kaufman** [comment]
@orenyodfat are you sure this is necessary with vote? I think the GP should send an executed event which will handle that.

**Comment by orenyodfat** [comment]
> @orenyodfat are you sure this is necessary with vote? I think the GP should send an executed event which will handle that.

Right .only with stake

**Comment by ben-kaufman** [comment]
@orenyodfat why did you close the PR?

**Comment by orenyodfat** [comment]
> @orenyodfat why did you close the PR?

mistake

File: src/mappings/genesisProtocol/GenesisProtocol.ts [hunk 6]
```diff
@@ -87,6 +90,19 @@ export function handleStake(event: Stake): void {
     return;
   }
 
+  let proposal = store.get(
```
Comment (orenyodfat) [inline src/mappings/genesisProtocol/GenesisProtocol.ts]: could you wrap this with a function (e.g stateUpdate) as it is called multiple times?

**Pull request merged**